### ISSUE TITLE
kubectl list pods in a deployment

### ISSUE THREAD

**Issue by dantman** [kubernetes/kubernetes#72794]
**What would you like to be added**:

Either an argument to `kubectl get pods` that allows you to list pods in a deployment; some way to get a selector string for `--selector` from a deployment in output_format'able way (`describe deployment` does not support -o and `get deployment` does not have selector in a format that `--selector` accepts); or a way to list pods in a deployment.yaml file.

**Why is this needed**:

I want a script that will check the status of pods in each of my deployments, things like whether the image is out of date. However right now it is impossible to list the pods in a deployment from a script without hardcoding selectors into the script.

- `kubectl get pods` accepts a `--selector` like `app=foo`
- `kubectl describe deployment foo` lists a selector in the same `app=foo` format
- However `kubectl describe` outputs in a human-readable format and does not support the `-o` output format option for scripts
- `kubectl get deployment foo -o=custom-columns=SELECTOR:.spec.selector` works but outputs the selector in a format like `map[matchLabels:map[app:foo]]`
- `kubetcl get -f deployment.yaml` does not list pods and you cannot use `kubectl get pods -f deployment.yaml`

So it seems like there is no way to list the pods in a deployment. Without reading a selector from a human-readable `describe` command and then manually inputting it into a `--selector`.

**Comment by dantman** [comment]
@kubernetes/sig-cli-feature-requests

**Comment by recollir** [comment]
What about `kubectl get deployment -o wide`? The last column returns selector-labels.

**Comment by dantman** [comment]
Is "wide" considered a "machine readable" format?

**Comment by liggitt** [comment]
> Is "wide" considered a "machine readable" format?

No, human readable default and wide column output is not guaranteed stable for scripting.

**Comment by dantman** [comment]
Ok, then we still need this feature in another output format.

/remove-lifecycle stale

**Comment by liggitt** [comment]
all the output formats for kubectl get display the result of a single API call. What this is requesting requires two API calls

**Comment by dantman** [comment]
@liggitt So listing pods by deployment directly is not possible. The other suggestion was outputting the selector in a machine readable way so that a script can just pass it to `--selector`, which should be within the realm of what an output format can do.

The problem is that yaml/json (machine readable) do not output selector in a format that can be passed to `--selector`; kubectl describe does not support output formats and is only human readable; and while `-o wide` outputs the selector it is not a machine readable output format.

**Comment by liggitt** [comment]
the selector is part of the deployment's scale subresource, which you can grab in a hacky way like this:

```
name=...
ns=...
selfLink=$(kubectl get deployment.apps "${name}" -n "${ns}" -o jsonpath={.metadata.selfLink})
selector=$(kubectl get --raw "${selfLink}/scale | jq -r .status.selector)
kubectl get pods -n "${ns}" --selector "${selector}"
```

**Comment by theAkito** [comment]
The workaround provided seems to already fail at trying to gather the selfLink. Is there no actual solution to this issue?

**Comment by callum-p** [comment]
Just updated to 1.20 today and the tool we use for deployments was using the selfLink field to monitor pods that are part of that specific deployment. selfLink was removed in 1.20. :(

**Comment by liggitt** [comment]
Updated my comment with a solution that doesn't use selfLink

**Comment by guillaumBrisard** [comment]
I think this issue needs to be reopened
It's still not straight forward with this tool

**Comment by javanthropus** [comment]
@liggitt's workaround doesn't work for DaemonSets or other resources that follow the same pattern.  This is more flexible:
```sh
NAMESPACE=...
RESOURCE=...    # e.g. deploy/foo or ds/bar
SELECTOR=$(
  kubectl get -n ${NAMESPACE} ${RESOURCE} -o json |
  jq -r '
    [
      (
        .spec.selector.matchLabels // empty |
        to_entries |
        map("\(.key)=\(.value)")
      ),
      (
        .spec.selector.matchExpressions // empty |
        map(
          if .operator == "Exists" then
            .key
          elif .operator == "DoesNotExist" then
            "!\(.key)"
          elif .operator == "In" then
            "\(.key) in (\(.values | join(", ")))"
          else
            "\(.key) notin (\(.values | join(", ")))"
          end
        )
      )
    ] |
    flatten |
    join(",")
  '
)
kubectl get pod -n ${NAMESPACE} -l ${SELECTOR}
```
This works with Deployments, DaemonSets, and Jobs and could be extended to also support CronJobs and other types that embed label selectors elsewhere in their definitions.
